[ROCm] add hipMAGMA support

jeffdaily opened the pull request:
Fixes #48831.  

- CI image is updated to build hipMAGMA from source and set env MAGMA_HOME.
- CMake is updated to separate different requirements for CUDA versus ROCm MAGMA.
- Some unit tests that become enabled with MAGMA are currently skipped for ROCm due to failures.  Fixing these failures will be follow-on work.

rraminen committed:
hipMAGMA enablement

jeffdaily committed:
For ROCm, add HAVE_HIP if USE_MAGMA is defined.

jeffdaily committed:
don't use CUDA_INCLUDE_DIRS if USE_ROCM and MAGMA_FOUND

rraminen committed:
Skipped failing PyTorch+hipMAGMA tests

rraminen committed:
Added hipMAGMA installation script for ubuntu

jeffdaily committed:
install hipMAGMA to /opt/rocm/magma

jeffdaily committed:
add gfx targets and rpath to hipMAGMA build

jeffdaily committed:
skip additional failing hipMAGMA tests

jeffdaily committed:
add env MAGMA_HOME to CI Dockerfile

jeffdaily commented:
ROCm CI puts us in a chicken and egg sort of situation.  This PR updates the CI image in addition to updating source files, tests, etc.  The ROCm CI Jenkins process uses a static image tag.  There's no clean way to actually test this PR; initial CI runs will use an older CI image that does not have MAGMA and will therefore appear to pass all tests.

jeffdaily commented:
Rebased against upstream master to try and fix unrelated CI failures.

malfet commented on .circleci/docker/common/install_rocm.sh:
```diff
@@ -44,9 +37,23 @@ install_ubuntu() {
+    cp make.inc-examples/make.inc.hip-mkl-gcc make.inc
+    echo 'LIBDIR += -L$(MKLROOT)/lib' >> make.inc
+    echo 'LIB += -Wl,--enable-new-dtags -Wl,--rpath,/opt/rocm/lib -Wl,--rpath,$(MKLROOT)/lib -Wl,--rpath,/opt/rocm/magma/lib' >> make.inc
+    echo 'DEVCCFLAGS += --amdgpu-target=gfx803 --amdgpu-target=gfx900 --amdgpu-target=gfx906 --amdgpu-target=gfx908' >> make.inc
```
Hmm, can we infer supported gpu targets from environment variable or something?

jeffdaily replied:
We don't currently have any way of querying the default set of targets for a particular ROCm release.  Today, every ROCm package is specifying their own set of gfx targets, though we do coordinate across projects to ensure they match.  [Example from HIP](https://github.com/ROCm-Developer-Tools/HIP/blob/2080cc113a2d767352b512b9d24c0620b6dee790/CMakeLists.txt#L370).

This PR matches that current list.  As we upgrade ROCm releases and prepare PRs for new ROCm versions, this list can change.

We're also discussing internally whether we can provide a hipMAGMA package alongside our other ROCm packages.  Such a package doesn't exist yet, so we're forced to build from source today.

malfet reviewed (approved):
LGTM, but can you please file issue and mention it in `@skipCUDAIfRocm`decorator, in order not to loose track about the reason why the test is skipped

jeffdaily commented:
The `@skipCUDAIfRocm` decorator doesn't allow a specific text description; adding issue as a comment instead.

jeffdaily committed:
add note in test/test_linalg.py pointing to github issue for ROCm magma

jeffdaily committed:
for ROCm, move magma install into function for both ubuntu and centos

jeffdaily commented:
@malfet I added 1 comment above the first skipped test.  Let me know if you had something else in mind.  Also moved the magma install into a separate function so it could be used for both ubuntu and centos images.

jeffdaily commented:
Rebased against `viable/strict`.
